compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeCoresAndObjectFifos.cpp

```diff
@@ -353,26 +450,179 @@ LogicalResult getUserTiles(
   return success();
 }
 
-/// Assign logical objectfifos to physical AIE tiles. This rewrite takes an
-/// iterative approach by matching logical objectfifos and only assigning tiles
-/// when linked through dma ops with other logical objectfifos which already
-/// have tiles assigned. If the linked logical objectfifos don't have tiles
-/// assigned yet, we will return a failure and give the linked logical
-/// objectfifos a chance to assign tiles before returning to this one.
+/// Insert `amdaie.logicalobjectfifo.access` operations which retrieve the
+/// memrefs from logical objectfifos and update the computational operations to
+/// operate on these local memrefs.
+LogicalResult insertLogicalObjectFifoAccess(ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp.getContext());
+
+  SmallVector<AMDAIE::CoreOp> coreOps;
+  moduleOp->walk([&](AMDAIE::CoreOp coreOp) { coreOps.push_back(coreOp); });
+
+  for (AMDAIE::CoreOp coreOp : coreOps) {
+    DenseMap<Value, std::tuple<AMDAIE::LogicalObjectFifoFromMemrefOp,
+                               AMDAIE::MemoryAccess>>
+        memrefToLogicalObjectFifo;
+    // First walk to collect consume/produce DMA accesses and map respective
+    // memrefs to logical objectifos.
+    coreOp->walk([&](Operation *op) {
+      // TODO(jornt): can we avoid produce/consume?
+      if (auto consumeOp = dyn_cast<AMDAIE::LogicalObjectFifoConsume>(op)) {
+        Value targetMemref =
+            consumeOp.getDmaCpyNdOp().getTargetObjectFifo().getMemref();
+        memrefToLogicalObjectFifo[targetMemref] =
+            std::make_pair(consumeOp.getDmaCpyNdOp().getTargetObjectFifo(),
+                           AMDAIE::MemoryAccess::Read);
+      } else if (auto produceOp =
+                     dyn_cast<AMDAIE::LogicalObjectFifoProduce>(op)) {
+        Value sourceMemref =
+            produceOp.getDmaCpyNdOp().getSourceObjectFifo().getMemref();
+        memrefToLogicalObjectFifo[sourceMemref] =
+            std::make_pair(produceOp.getDmaCpyNdOp().getSourceObjectFifo(),
+                           AMDAIE::MemoryAccess::Write);
+      }
+    });
+
+    WalkResult res = coreOp->walk([&](Operation *op) {
+      if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
+        for (auto &&[idx, operand] :
+             llvm::enumerate(linalgOp->getOpOperands())) {
+          if (memrefToLogicalObjectFifo.contains(operand.get())) {
+            rewriter.setInsertionPointToStart(coreOp.getBody());
+            std::tuple<AMDAIE::LogicalObjectFifoFromMemrefOp,
+                       AMDAIE::MemoryAccess>
+                value = memrefToLogicalObjectFifo[operand.get()];
+            rewriter.create<AMDAIE::LogicalObjectFifoAccessOp>(
+                rewriter.getUnknownLoc(), std::get<0>(value),
+                std::get<1>(value));
+            // TODO(jornt): Temporary, enable after access operations are used
```

Comment:
Temporary to avoid `LogicalObjectFifoAccessOp` uses for now. Will be uncommented in a follow-up, see https://github.com/nod-ai/iree-amd-aie/compare/jornt_cpp_pipeline?expand=1